tests/test_optimizer.py

```diff
@@ -694,3 +696,121 @@ def test_query_nested_fragments(db, gql_client: GraphQLTestClient):
+            "name": p.name,
+            "isDelayed": p.due_date < today,
+            "milestonesCount": len(ms),
+            "isSmall": len(ms) < 3,
+        }
+        expected.append(p_res)
+    for p in ProjectFactory.create_batch(
+        2,
+        due_date=today - datetime.timedelta(days=1),
+    ):
+        ms = MilestoneFactory.create_batch(2, project=p)
+        p_res: dict[str, Any] = {
+            "id": to_base64("ProjectType", p.id),
+            "name": p.name,
+            "isDelayed": p.due_date < today,
+            "milestonesCount": len(ms),
+            "isSmall": len(ms) < 3,
+        }
+        expected.append(p_res)
+
+    assert len(expected) == 4
+    for e in expected:
+        if DjangoOptimizerExtension.enabled.get():
+            with assert_num_queries(1):
+                res = gql_client.query(query, {"node_id": e["id"]})
+                assert res.data == {"project": e}
+        else:
+            # isDelayed and milestonesCount requires the optimizer to be turned on
+            res = gql_client.query(
+                query,
+                {"node_id": e["id"]},
+                asserts_errors=False,
+            )
+            assert res.errors
+
+
+@pytest.mark.django_db(transaction=True)
+def test_query_annotate_with_callable(db, gql_client: GraphQLTestClient):
+    query = """
+      query TestQuery ($node_id: GlobalID!) {
+        project (id: $node_id) {
+          id
+          name
+          milestones {
+            id
+            name
+            myBugsCount
+          }
+        }
+      }
+    """
+
+    user = UserFactory.create()
+    expected = []
+    for p in ProjectFactory.create_batch(2):
+        p_res: dict[str, Any] = {
+            "id": to_base64("ProjectType", p.id),
+            "name": p.name,
+            "milestones": [],
+        }
+        expected.append(p_res)
+        for m in MilestoneFactory.create_batch(2, project=p):
+            m_res: dict[str, Any] = {
+                "id": to_base64("MilestoneType", m.id),
+                "name": m.name,
+                "myBugsCount": 0,
+            }
+            p_res["milestones"].append(m_res)
+
+            # Those issues are not assigned to the user,
+            # thus they should not be counted
+            IssueFactory.create_batch(2, milestone=m, kind=Issue.Kind.BUG)
+            # Those issues are not bugs,
+            # thus they should not be counted
+            IssueFactory.create_batch(3, milestone=m, kind=Issue.Kind.FEATURE)
+            # Those issues are bugs assigned to the user,
+            # thus they will be counted
+            for i in IssueFactory.create_batch(4, milestone=m, kind=Issue.Kind.BUG):
+                Assignee.objects.create(user=user, issue=i)
+                m_res["myBugsCount"] += 1
+
+    assert len(expected) == 2
+    for e in expected:
+        with gql_client.login(user):
+            if DjangoOptimizerExtension.enabled.get():
+                with assert_num_queries(4):
```

Comment:
I haven't tried the code, but shouldn't this be 3 queries?

- 1 for the logged user
- 1 to retrieve the project 
- 1 to prefetch the milestones, with the `bugsCount` aggregate in it

Reply:
Here are the queries:

```
(Pdb) ctx.connection.queries[0]['sql']
'SELECT "django_session"."session_key", "django_session"."session_data", "django_session"."expire_date" FROM "django_session" WHERE ("django_session"."expire_date" > \'2023-09-28 17:19:11.112006\' AND "django_session"."session_key" = \'xepa46zwe72z635b6nby9jlvdl4zl7cl\') LIMIT 21'
(Pdb) ctx.connection.queries[1]['sql']
'SELECT "auth_user"."id", "auth_user"."password", "auth_user"."last_login", "auth_user"."is_superuser", "auth_user"."username", "auth_user"."first_name", "auth_user"."last_name", "auth_user"."email", "auth_user"."is_staff", "auth_user"."is_active", "auth_user"."date_joined" FROM "auth_user" WHERE "auth_user"."id" = 1 LIMIT 21'
(Pdb) ctx.connection.queries[2]['sql']
'SELECT "projects_project"."id", "projects_project"."name" FROM "projects_project" WHERE "projects_project"."id" = 1 ORDER BY "projects_project"."id" ASC LIMIT 1'
(Pdb) ctx.connection.queries[3]['sql']
'SELECT "projects_milestone"."id", "projects_milestone"."name", "projects_milestone"."project_id", COUNT("projects_issue"."id") FILTER (WHERE ("projects_assignee"."user_id" = 1 AND "projects_issue"."kind" = \'b\')) AS "_my_bugs_count" FROM "projects_milestone" LEFT OUTER JOIN "projects_issue" ON ("projects_milestone"."id" = "projects_issue"."milestone_id") LEFT OUTER JOIN "projects_assignee" ON ("projects_issue"."id" = "projects_assignee"."issue_id") WHERE "projects_milestone"."project_id" IN (1) GROUP BY "projects_milestone"."id", "projects_milestone"."name", "projects_milestone"."project_id"'
```

The second query is for the user, which happens also on `test_query_prefetch_with_callable`. So I guess this is fine?

Reply:
Oh, there's an extra query for the `session`

Yeah, it is fine :)